WalletWasabi/WabiSabi/Client/CoinJoinClient.cs

```diff
@@ -0,0 +1,191 @@
+		private SecureRandom SecureRandom { get; }
+		private CancellationTokenSource DisposeCts { get; } = new();
+		private Coin[] Coins { get; set; }
+
+		public CoinJoinClient(
+			Round round,
+			IArenaRequestHandler arenaRequestHandler,
+			BitcoinSecret bitcoinSecret,
+			IEnumerable<Coin> coins,
+			KeyManager keymanager)
+		{
+			Round = round;
+			ArenaRequestHandler = arenaRequestHandler;
+			BitcoinSecret = bitcoinSecret;
+			Keymanager = keymanager;
+			SecureRandom = new SecureRandom();
+			Coins = coins.ToArray();
+		}
+
+		protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+		{
+			try
+			{
+				// Register coins.
+				AliceClient[] aliceClients = await RegisterCoinsAsync().ConfigureAwait(false);
+
+				// Confirm coins.
+				await ConfirmConnectionsAsync(aliceClients, stoppingToken).ConfigureAwait(false);
+
+				Money[]? othersInputAmounts = null; // TODO: Get it from somewhere.
+
+				// Planning
+				var outputs = await PlanAmountsAsync(othersInputAmounts, stoppingToken).ConfigureAwait(false);
+
+				// Output registration.
+				await RegisterOutputsAsync(outputs).ConfigureAwait(false);
+
+				Transaction? unsignedCoinJoinTransaction = null; // TODO: Get it from somewhere.
+
+				// Send signature.
+				await SignTransactionAsync(aliceClients, unsignedCoinJoinTransaction).ConfigureAwait(false);
+			}
+			catch (Exception ex)
+			{
+			}
+		}
+
+		public async Task StartMixingCoinsAsync()
+		{
+			await StartAsync(DisposeCts.Token).ConfigureAwait(false);
+		}
+
+		private async Task<AliceClient[]> RegisterCoinsAsync()
+		{
+			var aliceArenaClient = new ArenaClient(
+				Round.AmountCredentialIssuerParameters,
+				Round.VsizeCredentialIssuerParameters,
+				AmountCredentialPool,
+				VsizeCredentialPool,
+				ArenaRequestHandler,
+				SecureRandom);
+
+			List<AliceClient> aliceClients = new();
+			try
+			{
+				foreach (var coin in Coins)
+				{
+					// Parallelize or Random delay?
+					aliceClients.Add(await AliceClient.CreateNewAsync(aliceArenaClient, coin, BitcoinSecret, Round.Id, Round.Hash, Round.FeeRate).ConfigureAwait(false));
+				}
+			}
+			catch (Exception)
+			{
+				foreach (var alice in aliceClients)
+				{
+					// Remove already registered Inputs.
+					try
+					{
+						await alice.RemoveInputAsync().ConfigureAwait(false);
+					}
+					catch (Exception)
+					{
+						// Log?
+					}
+				}
+				throw;
+			}
+
+			return aliceClients.ToArray();
+		}
+
+		private async Task ConfirmConnectionsAsync(AliceClient[] aliceClients, CancellationToken stoppingToken)
+		{
+			Task[] confirmTasks = aliceClients.Select(aliceClient => aliceClient.ConfirmConnectionAsync(TimeSpan.FromSeconds(5), stoppingToken)).ToArray();
+
+			await Task.WhenAll(confirmTasks).ConfigureAwait(false);
+
+			var exceptions = confirmTasks.Where(t => t.IsFaulted && t.Exception is { }).Select(t => t.Exception);
+			if (exceptions.Any())
+			{
+				// Error! Try to de-register inputs?
```

Comment:
If there are any errors here, we are out of the round. I would not tangle into any fallback scenarios, it can be extra complicated and IMO slow down the process. This is background mixing, if it cannot be done this round we will do it in the next.

Reply:
Confirmed inputs can't be de-registered though, they are committed to participating from the coordinator side. see also previous comment about when input registration fails. Simply ignoring is just as much of a privacy leak as deregistering.

Reply:
Done